mesh/mesh_readers.cpp

```diff
@@ -886,36 +886,299 @@ void Mesh::ReadInlineMesh(std::istream &input, int generate_edges)
+                                    std::vector<Element*> &elements_1D,
+                                    std::vector<Element*> &elements_2D,
+                                    std::vector<Element*> &elements_3D)
 {
-   string buff;
-   double version;
-   int binary, dsize;
-   input >> version >> binary >> dsize;
-   if (version < 2.2)
+   if (!elements_3D.empty())
    {
-      MFEM_ABORT("Gmsh file version < 2.2");
+      Dim = 3;
+      NumOfElements = elements_3D.size();
+      elements.SetSize(NumOfElements);
+      for (int el = 0; el < NumOfElements; ++el)
+      {
+         elements[el] = elements_3D[el];
+      }
+      NumOfBdrElements = elements_2D.size();
+      boundary.SetSize(NumOfBdrElements);
+      for (int el = 0; el < NumOfBdrElements; ++el)
+      {
+         boundary[el] = elements_2D[el];
+      }
+      // discard other elements
+      for (size_t el = 0; el < elements_1D.size(); ++el)
+      {
+         delete elements_1D[el];
+      }
+      for (size_t el = 0; el < elements_0D.size(); ++el)
+      {
+         delete elements_0D[el];
+      }
    }
-   if (dsize != sizeof(double))
+   else if (!elements_2D.empty())
    {
-      MFEM_ABORT("Gmsh file : dsize != sizeof(double)");
+      Dim = 2;
+      NumOfElements = elements_2D.size();
+      elements.SetSize(NumOfElements);
+      for (int el = 0; el < NumOfElements; ++el)
+      {
+         elements[el] = elements_2D[el];
+      }
+      NumOfBdrElements = elements_1D.size();
+      boundary.SetSize(NumOfBdrElements);
+      for (int el = 0; el < NumOfBdrElements; ++el)
+      {
+         boundary[el] = elements_1D[el];
+      }
+      // discard other elements
+      for (size_t el = 0; el < elements_0D.size(); ++el)
+      {
+         delete elements_0D[el];
+      }
    }
-   getline(input, buff);
-   // There is a number 1 in binary format
-   if (binary)
+   else if (!elements_1D.empty())
    {
-      int one;
-      input.read(reinterpret_cast<char*>(&one), sizeof(one));
-      if (one != 1)
+      Dim = 1;
+      NumOfElements = elements_1D.size();
+      elements.SetSize(NumOfElements);
+      for (int el = 0; el < NumOfElements; ++el)
       {
-         MFEM_ABORT("Gmsh file : wrong binary format");
+         elements[el] = elements_1D[el];
+      }
+      NumOfBdrElements = elements_0D.size();
+      boundary.SetSize(NumOfBdrElements);
+      for (int el = 0; el < NumOfBdrElements; ++el)
+      {
+         boundary[el] = elements_0D[el];
       }
    }
+   else
+   {
+      MFEM_ABORT("Gmsh file : no elements found");
+      return;
+   }
+}
 
+void Mesh::ReadGmshV4(std::istream &input, int binary)
+{
+
+   string buff;
    // A map between a serial number of the vertex and its number in the file
    // (there may be gaps in the numbering, and also Gmsh enumerates vertices
    // starting from 1, not 0)
-   map<int, int> vertices_map;
+   map<std::size_t, std::size_t> vertices_map;
```

Comment:
Do size_t's really make more sense here?

Reply:
Not in the grand scheme. Only in the sense that the type returned by gmesh for it's serial number is unsigned long. size_t is arguably more efficient as natural memory fetch size.  But, are int's better (are we expecting negative indices, ;) But, it is inconsistent with the rest of the types expressed in the other readers, and throughout the mesh code. So, I was thinking this should be changed back too.

Reply:
Also, if you can spare the cycles, go ahead and make those changes.